Add RESTful API Guide (#3414)
ballerina-platform/ballerina-dev-website: learn/guides/developing-a-restful-api-with-ballerina.md

---
layout: developing-a-restful-api-with-ballerina
title: Developing a RESTful API with Ballerina
permalink: /learn/developing-a-restful-api-with-ballerina/
description: This simple guide helps you understand the basics of Ballerina constructs which allow you to write RESTful APIs. 
active: developing-a-restful-api-with-ballerina
intro: This simple guide helps you understand the basics of Ballerina constructs which allow you to write RESTful APIs. 
redirect_from:

>   - /learn/developing-a-restful-api-with-ballerina

Reviewer: Do we need the redirect from here?

Reviewer 2: We need to add the entry without the slash.

---

> This simple guide helps you understand the basics of Ballerina constructs which allow you to write RESTful APIs. 

Reviewer 2: ```suggestion
This simple guide helps you understand the basics of Ballerina constructs which allow you to write RESTful APIs. 
```
Please see the review comments in the GrpahQL guide PR: https://github.com/ballerina-platform/ballerina-dev-website/pull/3369

> Due to the batteries included nature of Ballerin language there is no need to add any third party libraries to implement the RESTful API. The Ballerina standard library itself is adequate. In this API you will be writing a simple CRUD like RESTful service. 

Reviewer 2: ```suggestion
Due to the batteries included nature of the  Ballerin language, there is no need to add any third-party libraries to implement the RESTful API. The Ballerina standard library itself is adequate. In this API, you will be writing a simple CRUD-like RESTful service. 
```

> To get the best out of the guide it is better to have some idea about Ballerina.

Reviewer 2: ```suggestion
To get the best out of the guide it is better to have some idea about Ballerina.
```
Do we need to mention this?

In this tutorial, you will be creating a RESTful API with two endpoints. The sample is built around a set of covid19 data.

> The tutorial includes the following steps,

Reviewer 2: ```suggestion
The tutorial includes the following steps,
```
Do we need this? The left navigation will show the sub topics.

Author: We thought leaving it for now.

1. Design the two endpoints 
2. Create the covid19 dataset 
3. Write a resource to get all the covid19 data
4. Write a resource to add covid19 data by ISO code
5. Write a resource to get filtered covid19 data by ISO code

## Prerequisite

> Following two prerequisites are needed for this tutorial. 

Reviewer 2: ```suggestion
The following two prerequisites are needed for this tutorial. 
```

1. [Ballerina Swan Lake](https://ballerina.io/downloads/)
2. [VSCode](https://code.visualstudio.com/download) with [Ballerina extension](https://marketplace.visualstudio.com/items?itemName=WSO2.Ballerina)

## Design the Two Endpoints 

> The first endpoint is about getting data from the service as well as adding data to the service. Therefore, the service should handle both HTTP GET and POST requests. The GET request is to get data whereas the POST request is to add data. In the case of the GET request the response should be 200 OK whereas in the case of POST request the response should be 201 created.

Reviewer 2: ```suggestion
The first endpoint is about getting data from the service as well as adding data to the service. Therefore, the service should handle both HTTP GET and POST requests. The GET request is to get data whereas the POST request is to add data. In the case of the GET request, the response should be 200 OK whereas, in the case of the POST request, the response should be 201 created.
```

> The second endpoint is about getting filtered data from the service. The data is filtered by the ISO code. Therefore, the second service accepts the ISO code as part of the URL and then responds with 200 OK status code. In the event of an error the relevant error is sent back to the client.

Reviewer 2: ```suggestion
The second endpoint is about getting data filtered from the service. The data is filtered by the ISO code. Therefore, the second service accepts the ISO code as part of the URL and then responds back with a 200 OK status code. In the event of an error, the relevant error is sent back to the client.
```

> Following is the URL for each endpoint respectively.

Reviewer 2: ```suggestion
The following is the URL for each endpoint respectively.
```

> 1. /covid/status/countries

Reviewer 2: ```suggestion
1. `/covid/status/countries`
```

> 2. /covid/status/countries/{iso_code}

Reviewer 2: ```suggestion
2. `/covid/status/countries/{iso_code}`
```

## Create a Package for Your Code

Ballerina uses packages to group code. In this case, a package with the default module is created by executing the following command.

`bal new covid19 -t service`

> This creates a folder named covid19 along with a sample code for a service. Move to the new folder and execute the following command to start VSCode.

Reviewer 2: ```suggestion
This creates a folder named covid19 along with a sample code for a service. Navigateto the new folder, and execute the following command to start VSCode.
```

`code .`

> `Ballerina.toml` is the file which actually makes the folder a Ballerina package. It also contains a test directory to include tests for the service. But for the sake of simplicity we will ignore it in this tutorial. You can just go through the sample to get a look and feel about Ballerina services. However, we will be starting with a blank page. Hence, before you start you can delete the entire code or edit it if you wish.

Reviewer 2: ```suggestion
`Ballerina.toml` is the file, which makes the folder a Ballerina package. It also contains a test directory to include tests for the service. However, for the sake of simplicity, you will ignore it in this tutorial. You can just go through the sample to get a look and feel about Ballerina services. However, we will be starting with a blank page. Hence, before you start, you can delete the entire code or edit it if you wish.
```
Change all occurrences of `we` to `you`.

## Create the Covid19 Dataset 

> To keep things simple a in memory dataset is used with three entries. Ballerina tables are used to store data. Each entry in the table is represented by a Ballerina record. Following is the definition of the record and the declaration of the table.

Reviewer 2: ```suggestion
To keep things simple, an in-memory dataset is used with three entries. Ballerina tables are used to store data. Each entry in the table is represented by a Ballerina record. The following is the definition of the record and the declaration of the table.
```

```ballerina
public type CovidEntry record {|
    readonly string iso_code;
    string country;
    decimal cases;
    decimal deaths;
    decimal recovered;
    decimal active;
|};
public final table<CovidEntry> key(iso_code) covidTable = table [
    { iso_code: "AFG", country: "Afghanistan", cases: 159303, deaths: 7386, recovered: 146084, active: 5833 },
    { iso_code: "SL", country: "Sri Lanka", cases: 598536, deaths: 15243, recovered: 568637, active: 14656 },
    { iso_code: "US", country: "USA", cases: 69808350, deaths: 880976, recovered: 43892277, active: 25035097 }
];
```

> ## Write a Resource to Get All the Covid19 Data

Reviewer 2: ```suggestion
## Writing a Resource to Get All the Covid19 Data
```
The `ing` form should be used in titles for consistency.

> As mentioned earlier, the first endpoint has two parts: getting data as well as adding data. In this section, the focus is on getting data. 

Reviewer 2: ```suggestion
As mentioned earlier, the first endpoint has two parts: getting data as well as adding data. In this section, the focus is on getting data. 
```
Shall we add a topic to highlight this?

> Ballerina resources can only be inside a service. Therefore, first a service needs to be created. If you have noticed, both endpoints have a common URL segment. When creating the service, the common URL segment can be moved to service level as the base path. 

Reviewer 2: ```suggestion
Ballerina resources can only reside inside a service. Therefore, first, a service needs to be created. If you have noticed, both endpoints have a common URL segment. When creating the service, the common URL segment can be moved to the service level as the base path. 
```

> Each service is associated with a `http:Listener`, it is the Ballerina abstraction which deals with network level details such as host, port, SSL, etc. 

Reviewer 2: ```suggestion
Each service is associated with an `http:Listener`. It is the Ballerina abstraction, which deals with the network-level details such as host, port, SSL, etc. 
```

```ballerina
service /covid/status on new http:Listener(9000) {
}   
```

> Adding the first endpoint to the service.

Reviewer 2: ```suggestion
Adding the first endpoint to the service.
```
Shall we add these as numbered bullet points of this is a sequence of steps?

```ballerina
service /covid/status on new http:Listener(9000){
	resource function get countries() returns CovidEntry[] {
        return covidTable.toArray();
    }
}
```

> Unlike normal functions resource functions can have accessors. In this case accessor is set to `get`, which means only HTTP GET requests could hit this resource. 

Reviewer 2: ```suggestion
Unlike normal functions, resource functions can have accessors. In this case, the accessor is set to `get`, which means only HTTP GET requests could hit this resource. 
```

> Ballerina automatically serializes Ballerina records as JSON and sends them over the wire. The default status code HTTP responses are 200 OK.

Reviewer 2: ```suggestion
Ballerina automatically serializes Ballerina records as JSON and sends them over the wire. The default status code for HTTP responses are 200 OK.
```

## Write a Resource to Add Covid19 Data by ISO Code

> The second part of the first endpoint is about adding or appending new data to the dataset. Following is the second resource associated with the first endpoint.

Reviewer 2: ```suggestion
The second part of the first endpoint is about adding or appending new data to the dataset. The following is the second resource associated with the first endpoint.
```

```ballerina
resource function post countries(@http:Payload CovidEntry covidEntry) 
                                    returns CreatedCovidEntry|ConflictingIsoCodeError {
    if covidTable.hasKey(covidEntry.iso_code) {
        return <ConflictingIsoCodeError>{ 
            body : { 
                errmsg : string `Conflicting ISO Code: ${covidEntry.iso_code}`
            } 
        };
    }
    covidTable.add(covidEntry); 
    return <CreatedCovidEntry>{ body: covidEntry }; 
}
```

> Copying this straightway results in an error which is expected as the `ConflictingIsoCodeError` type is not defined yet.

Reviewer 2: ```suggestion
Copying this straightway results in an error, which is expected as the `ConflictingIsoCodeError` type is not defined yet.
```

> This resource is a bit more interesting than the second resource. To begin with there is a resource argument named `covidEntry` annotated with `@http:Payload`. This basically means the resource is expecting a payload with type `CovideEntry`. 

Reviewer 2: ```suggestion
This resource is a bit more interesting than the second resource. There is a resource argument named `covidEntry` annotated with `@http:Payload`. This basically means the resource is expecting a payload with the `CovideEntry` type. 
```

> Finally, as the return values there are two types of records `CovidEntry` and `ConflictingIsoCodeError`. The former is the same record that was used in the first resource. The other is a bit different and the following is the definition of `ConflictingIsoCodeError`.

Reviewer 2: ```suggestion
Finally, there are two types of records `CovidEntry` and `ConflictingIsoCodeError` as the return values. The former is the same record that was used in the first resource. The other is a bit different and the following is the definition of the `ConflictingIsoCodeError`.
```

```ballerina
public type ConflictingIsoCodeError record {|
    *http:Conflict;
    ErrorMsg body;
|};
public type ErrorMsg record {|
    string errmsg;
|};
```

> As you can see, there is something new here. `*http:Conflict` is the Ballerina way of saying one type is a subtype another. In this case `ConflictingIsoCodeError` is a subtype of `*http:Conflict`.Ballerina has defined a set of types for each HTTP status code. This allows users to write services in a type oriented way which is in turn helpful when it comes to tooling and generating OpenAPI specification for HTTP services. Returning this record results in HTTP 409 response with JSON payload. 

Reviewer 2: ```suggestion
As you can see, there is something new here. The `*http:Conflict` is the Ballerina way of saying one type is a subtype of another. In this case, the `ConflictingIsoCodeError` is a subtype of the `*http:Conflict`. Ballerina has defined a set of types for each HTTP status code. This allows you to write services in a type-oriented way, which is, in turn, helpful when it comes to tooling and generating OpenAPI specifications for HTTP services. Returning this record results in an HTTP 409 response with a JSON payload. 
```

> The body of the response is of type `ErrorMsg` which simply has a string field named `errmsg`. Based on the need users can have any data type for their response body.

Reviewer 2: ```suggestion
The body of the response is of the `ErrorMsg` type, which simply has a string field named `errmsg`. Based on the need, you can have any data type for their response body.
```

## Write a Resource to Get Filtered Covid19 Data by ISO Code

> This resource is a bit more different than the first two resources. As you already know resource functions have accessors. In addition, it also supports hierarchical paths. Making it ideal for implementing RESTful APIs. Hierarchical paths can have path params. In this case iso_code is used as the path param which intern becomes a string variable. 

Reviewer 2: ```suggestion
This resource is a bit more different than the first two resources. Resource functions have accessors. In addition, it also supports hierarchical paths making it ideal for implementing RESTful APIs. Hierarchical paths can have path params. In this case, `iso_code` is used as the path param, which in turn, becomes a string variable. 
```

```ballerina
resource function get countries/[string iso_code]() returns CovidEntry|InvalidIsoCodeError {
    CovidEntry? covidEntry = covidTable[iso_code];
    if covidEntry is () {
        return { 
            body: { 
                errmsg: string `Invalid ISO Code: ${iso_code}`
            }
        };
    }
    return covidEntry;
}
```

> As in the previous example, this resource also includes its own return types. But the basic principle behind them is as the previous example. Following is the type definition of `InvalidIsoCodeError`.

Reviewer 2: ```suggestion
As in the previous example, this resource also includes its own return types. However, the basic principle behind them is as the previous example. The following is the type definition of the `InvalidIsoCodeError`.
```

```ballerina
public type InvalidIsoCodeError record {|
    *http:NotFound;
    ErrorMsg body;
|};
```
## The Complete Code

> It is always a good practice to document your interfaces. However, this example has omitted documentation for brevity. But any production ready API interface must include API documentation. 

Reviewer 2: ```suggestion
It is always a good practice to document your interfaces. However, this example has omitted documentation for brevity. Nevertheless, any production-ready API interface must include API documentation. 
```

You can run the service by executing the following command inside the `covid19` folder. 

`bal run`

> > **Note**: Console should have warning logs related to isolatedness of resources. It is a built-in service concurrency safety feature in Ballerina. 

Reviewer 2: ```suggestion
> **Note**: The console should have warning logs related to the isolatedness of resources. It is a built-in service concurrency safety feature of Ballerina. 
```

> You can also try generating OpenAPI specification for the written service by executing the following command which creates a `yaml` file in the current folder.

Reviewer 2: ```suggestion
You can also try generating an OpenAPI specification for the written service by executing the following command, which creates a `yaml` file in the current folder.
```
